### ISSUE TITLE
IDX10511: Signature validation failed

### ISSUE THREAD

**Issue by JoelPagliuca** [Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2#209]
### This issue is for a: (mark with an `x`)
```
- [x] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### The issue was found for the following scenario:

Please add an 'x' for the scenario(s) where you found an issue

1. Web app that signs in users
   1. [ ] with a work and school account in your organization: [1-WebApp-OIDC/1-1-MyOrg](../blob/master/1-WebApp-OIDC/1-1-MyOrg)
   1. [ ] with any work and school account: [/1-WebApp-OIDC/1-2-AnyOrg](../blob/master/1-WebApp-OIDC/1-2-AnyOrg)
   1. [ ] with any work or school account or Microsoft personal account: [1-WebApp-OIDC/1-3-AnyOrgOrPersonal](../blob/master/1-WebApp-OIDC/1-3-AnyOrgOrPersonal)
   1. [ ] with users in National or sovereign clouds [1-WebApp-OIDC/1-4-Sovereign](../blob/master/1-WebApp-OIDC/1-4-Sovereign)
   1. [ ] with B2C users [1-WebApp-OIDC/1-5-B2C](../blob/master/1-WebApp-OIDC/1-5-B2C)
1. Web app that calls Microsoft Graph
   1. [ ] Calling graph with the Microsoft Graph SDK: [2-WebApp-graph-user/2-1-Call-MSGraph](../blob/master/2-WebApp-graph-user/2-1-Call-MSGraph)
   1. [ ] With specific token caches: [2-WebApp-graph-user/2-2-TokenCache](../blob/master/2-WebApp-graph-user/2-2-TokenCache)
   1. [ ] Calling Microsoft Graph in national clouds: [2-WebApp-graph-user/2-4-Sovereign-Call-MSGraph](../blob/master/2-WebApp-graph-user/2-4-Sovereign-Call-MSGraph)
1. [ ] Web app calling several APIs [3-WebApp-multi-APIs](../blob/master/3-WebApp-multi-APIs)
1. [x] Web app calling your own Web API [4-WebApp-your-API](../blob/master/4-WebApp-your-API)
1. Web app restricting users
   1. [ ] by Roles: [5-WebApp-AuthZ/5-1-Roles](../blob/master/5-WebApp-AuthZ/5-1-Roles)
   1. [ ] by Groups: [5-WebApp-AuthZ/5-2-Groups](../blob/master/5-WebApp-AuthZ/5-2-Groups)
1. [ ] Deployment to Azure
1. [ ] Other (please describe)

### Repro-ing the issue

**Repro steps**

I'm building a protected Web API (by the definition given here an API that will use AzureAD as source of Identity and authentication).

I followed the guide in `4-WebApp-your-API`
* I have the `Microsoft.Identity.Web` directory copied into my current project
* I have the `services.AddProtectedWebApi(Configuration);` line in my code with a properly configured `AzureAd` section of an `appsettings.json`
*  I have a `app.UseAuthentication();` line in my `Configure()` method and `[Authorize]` decorator on my API Controller
I'm now testing my API using Postman. When requesting an OAuth2.0 token with Implicit Flow
* Using this Auth URL: https://login.microsoftonline.com/`tenant id`/oauth2/v2.0/authorize
And now I use that token to as a Bearer token to call my API.

**Expected behavior**
200 response with data from my API because I expect the JWT middleware provided by `AddProtectedWebApi` to handle the access_token given through this flow as described by this [ms doc](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-implicit-grant-flow)

**Actual behavior**
dotnet exception: `IDX10511: Signature validation failed`

**Possible Solution**
I feel like this is a problem with the custom JWTs that Azure is using. Maybe I'm missing something in my assumptions or testing? Is there a better alternative of getting an `access_token` that might solve my problem?

**Additional context/ Error codes / Screenshots**

### Any log messages given by the failure
```
Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler[7]
      AzureADJwtBearer was not authenticated. Failure message: IDX10511: Signature validation failed. Keys tried: 'Microsoft.IdentityModel.Tokens.X509SecurityKey , KeyId: BB8CeFVqyaGrGNuehJIiL4dfjzw
      '. 
      kid: 'BB8CeFVqyaGrGNuehJIiL4dfjzw'. 
      Exceptions caught:
       ''.
      token: <a correctly decoded token>.
```

- You can enable Middleware diagnostics by uncommenting the following [lines](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/418e4880ce3307befb25c7af600a886560cadcaa/Microsoft.Identity.Web/StartupHelpers.cs#L81-L83)
- You can enable personally identifiable information in your exceptions to get more information in the open id connect middleware see [Seeing [PII is hidden] in log messages](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/PII) 
- Logging for MSAL.NET is described at [Loggin in MSAL.NET](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-logging#logging-in-msalnet)

### OS and Version?
> macOS

### Versions
> dotnet core 2.2.105

### Attempting to troubleshooting yourself:

- did you go through the README.md in the folder where you found the issue?
- did you go through the documentation:
  - [Scenario: Web app that signs in users](https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-web-app-sign-user-overview)
  - [Scenario: Web app that calls web APIs](https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-web-app-call-api-overview)
  
### Mention any other details that might be useful

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

**Comment by TiagoBrenck** [comment]
Hi @JoelPagliuca , when you get an access token for Graph, it can only be used to consume Graph API. If you have your own web api, you must get another access token (issued to your web api) and send it as bearer, not the Graph one.

I like to say that access tokens are like visas to enter a country. If you have a US visa, you cannot use it to enter Canada for instance. The same goes to web apis. If you have an access token for Graph, you cannot use it to consume another web api other than Graph.

**Comment by JoelPagliuca** [comment]
That's a great analogy @TiagoBrenck thank-you!
How am I supposed to be generating an `access_token` using this library? Then documentation is a bit hard to follow at the moment and I think I've gone in circles a few times 😅.
So far I had the understanding my SPA was supposed to give my API an `access_token` from login.microsoft in the Auth header and the `services.AddProtectedWebApi` was supposed to take care of the rest

**Comment by TiagoBrenck** [comment]
Have you complete [this step](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/master/4-WebApp-your-API#register-the-service-app-todolistservice-aspnetcore-webapi) for registering your web api?

**Comment by JoelPagliuca** [comment]
Aha! I had not, does my SPA now become a client of my Web API? What does the flow now look like for logging into my application with an Implicit flow using Azure as the Identity provider?

**Comment by JoelPagliuca** [comment]
Or not even specifically an Implicit flow, myself and the front-end developer working with me are just keen to access this API I'm building with a JWT

**Comment by TiagoBrenck** [comment]
So, once you have created the WebApi application in azure portal, and followed the **Expose an API** step, you will have the scope that your SPA will use when requesting an Access Token.

I don't know how your code is set, but I assume you were able to get a token from Graph. So in your code, try replacing the graph scope that you were using with the one you ve just created on your web api.

**Comment by JoelPagliuca** [comment]
So it's looking like I get a valid JWT `access_token` now for my Web API, thank you so much for this help! 

I'll try and see if I can raise a PR to suggest changes in the docs to help others find this advice quicker. It looks like the best flow to get here is by adding the App Registration and following the linked examples to this repository. I originally tried using dotnet core [documentation](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/identity?view=aspnetcore-2.2&tabs=netcore-cli) to add authentication to my API (as I've previously used the same documentation to implement Google Auth in a different project).

Out of curiosity, where do you recommend I start reading if I want to now get the AD groups attached to the user within the application?

**Comment by TiagoBrenck** [comment]
That is awesome news!

For groups, you could check [the step 5 in this tutorial](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/master/5-WebApp-AuthZ/5-2-Groups).

**Comment by JoelPagliuca** [comment]
Thanks mate, I'll have a crack at that shortly, I'll close this issue :)
